src/DataStructures/LinkedMessageQueue.hpp

```diff
@@ -0,0 +1,178 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <functional>
+#include <optional>
+#include <ostream>
+#include <pup.h>
+#include <unordered_map>
+#include <utility>
+
+#include "Parallel/PupStlCpp17.hpp"
+#include "Utilities/ErrorHandling/Assert.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TaggedTuple.hpp"
+
+/// \ingroup DataStructuresGroup
+/// An identifier for an element in a sequence
+///
+/// The struct contains an ID and the ID for the previous element of
+/// the sequence, if any.  This allows the sequence to be
+/// reconstructed even when elements are received out-of-order.
+///
+/// \see LinkedMessageQueue
+template <typename Id>
+struct LinkedMessageId {
+  Id id;
+  std::optional<Id> previous;
+
+  // NOLINTNEXTLINE(google-runtime-references)
+  void pup(PUP::er& p) noexcept {
+    p | id;
+    p | previous;
+  }
+};
+
+template <typename Id>
+bool operator==(const LinkedMessageId<Id>& a,
+                const LinkedMessageId<Id>& b) noexcept {
+  return a.id == b.id and a.previous == b.previous;
+}
+
+template <typename Id>
+bool operator!=(const LinkedMessageId<Id>& a,
+                const LinkedMessageId<Id>& b) noexcept {
+  return not(a == b);
+}
+
+template <typename Id>
+std::ostream& operator<<(std::ostream& s,
+                         const LinkedMessageId<Id>& id) noexcept {
+  return s << id.id << " (" << id.previous << ")";
+}
+
+namespace std {
+template <typename Id>
+struct hash<LinkedMessageId<Id>> {
+  size_t operator()(const LinkedMessageId<Id>& id) const noexcept {
+    // For normal use, any particular .id should only appear with one
+    // .previous, and vice versa, so hashing only one is fine.
+    return std::hash<Id>{}(id.id);
+  }
+};
+}  // namespace std
+
+/// \cond
+template <typename Id, typename QueueTags>
+class LinkedMessageQueue;
+/// \endcond
+
+/// \ingroup DataStructuresGroup
+/// A collection of queues of asynchronously received data
+///
+/// This class is designed to collect asynchronously received
+/// messages, possibly from multiple sources, until sufficient data
+/// has been received for processing.  Each tag in \p QueueTags
+/// defines a queue of messages of type `Tag::type`.
+template <typename Id, typename... QueueTags>
+class LinkedMessageQueue<Id, tmpl::list<QueueTags...>> {
+  static_assert(sizeof...(QueueTags) > 0,
+                "Must have at least one message source.");
+
+ public:
+  using IdType = Id;
+
+  /// Insert data into a given queue at a given ID.  All queues must
+  /// receive data with the same collection of \p id_and_previous, but
+  /// are not required to receive them in the same order.
+  template <typename Tag>
+  void insert(const LinkedMessageId<Id>& id_and_previous,
+              typename Tag::type message) noexcept;
+
+  /// The next ID in the received sequence, if all queues have
+  /// received messages at that ID.
+  std::optional<Id> ready_id() const noexcept;
```

Comment:
(optional) Maybe rename to `next_ready_id()` or `earliest_ready_id()` to clarify that if there are multiple ids that are ready, this will grab the "earliest" one in the queue. (You have a test where this happens I believe)